### ISSUE TITLE
metamorphic (race): failure from 5/19

### ISSUE THREAD

**Issue by nicktrav** [cockroachdb/pebble#1698]
Found via a metamorphic test run ([logs](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/5202092?buildTab=log&focusLine=5281&linesState=4494)).

```console
===== SEED =====
1652935592018725775
===== ERR =====
exit status 66
===== OUT =====
==================
WARNING: DATA RACE
Write at 0x00c000233b18 by goroutine 25:
  runtime.slicecopy()
      GOROOT/src/runtime/slice.go:284 +0x0
  github.com/cockroachdb/pebble/sstable.(*blockIter).readEntry()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/block.go:416 +0xb7e
  github.com/cockroachdb/pebble/sstable.(*blockIter).First()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/block.go:793 +0xfd
  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).firstInternal()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:791 +0x111
  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).First()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:768 +0x64
  github.com/cockroachdb/pebble.(*levelIter).skipEmptyFileForward()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:621 +0x74
  github.com/cockroachdb/pebble.(*levelIter).Next()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:560 +0x3a8
  github.com/cockroachdb/pebble.(*simpleMergingIter).step()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_checker.go:226 +0x2431
  github.com/cockroachdb/pebble.checkLevelsInternal()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_checker.go:660 +0x1629
  github.com/cockroachdb/pebble.(*DB).CheckLevels()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_checker.go:576 +0x264
  github.com/cockroachdb/pebble.DebugCheckLevels()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/options.go:736 +0x32
  github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func1.1()
      github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/test.go:58 +0x33
  github.com/cockroachdb/pebble/internal/metamorphic.withRetries()
      github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/retryable.go:18 +0x2e
  github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func1()
      github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/test.go:57 +0x4a
  github.com/cockroachdb/pebble.(*DB).updateReadStateLocked()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/read_state.go:105 +0x33b
  github.com/cockroachdb/pebble.(*DB).compact1()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1916 +0xc0c
  github.com/cockroachdb/pebble.(*DB).compact.func1()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1840 +0xe4
  runtime/pprof.Do()
      GOROOT/src/runtime/pprof/runtime.go:40 +0x125
  github.com/cockroachdb/pebble.(*DB).compact()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1837 +0xb5
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker·dwrap·13()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x58

Previous read at 0x00c000233b1e by goroutine 7:
  [failed to restore the stack]

Goroutine 25 (running) created at:
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x1044
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompaction()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1530 +0x1a9
  github.com/cockroachdb/pebble.(*DB).compact.func1()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1847 +0x19d
  runtime/pprof.Do()
      GOROOT/src/runtime/pprof/runtime.go:40 +0x125
  github.com/cockroachdb/pebble.(*DB).compact()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1837 +0xb5
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker·dwrap·13()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x58

Goroutine 7 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      GOROOT/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      GOROOT/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1504 +0x9d1
  main.main()
      main/bazel-out/k8-dbg/bin/external/com_github_cockroachdb_pebble/internal/metamorphic/metamorphic_test_/testmain.go:131 +0x5c4
==================
```

SHA: `9bd431430a38`.

_NOTE: I filled this in manually, as we didn't get an automated issue._

**Comment by nicktrav** [comment]
cc: @bananabrick - I think you're already digging into this one.

**Comment by bananabrick** [comment]
Got a repro locally:

```
 WARNING: DATA RACE
        Write at 0x00c0002b1218 by goroutine 96:
          runtime.slicecopy()
              /snap/go/9760/src/runtime/slice.go:295 +0x0
          github.com/cockroachdb/pebble/sstable.(*blockIter).readEntry()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/sstable/block.go:416 +0xb78
          github.com/cockroachdb/pebble/sstable.(*blockIter).First()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/sstable/block.go:793 +0x144
          github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).firstInternal()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/sstable/reader.go:791 +0x111
          github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).First()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/sstable/reader.go:768 +0x64
          github.com/cockroachdb/pebble.(*levelIter).Next()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/level_iter.go:546 +0x215
          github.com/cockroachdb/pebble.(*simpleMergingIter).step()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/level_checker.go:226 +0x2410
          github.com/cockroachdb/pebble.checkLevelsInternal()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/level_checker.go:660 +0x1609
          github.com/cockroachdb/pebble.(*DB).CheckLevels()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/level_checker.go:576 +0x244
          github.com/cockroachdb/pebble.DebugCheckLevels()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/options.go:736 +0x32
          github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func1.1()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/test.go:58 +0x33
          github.com/cockroachdb/pebble/internal/metamorphic.withRetries()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/retryable.go:18 +0x2e
          github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func1()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/test.go:57 +0x4a
          github.com/cockroachdb/pebble.(*DB).updateReadStateLocked()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/read_state.go:105 +0x33b
          github.com/cockroachdb/pebble.(*DB).compact1()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/compaction.go:1916 +0xd4c
          github.com/cockroachdb/pebble.(*DB).compact.func1()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/compaction.go:1840 +0xd2
          runtime/pprof.Do()
/snap/go/9760/src/runtime/pprof/runtime.go:40 +0x122
          github.com/cockroachdb/pebble.(*DB).compact()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/compaction.go:1837 +0xb5
          github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker.func4()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/compaction.go:1631 +0x49


Previous read at 0x00c0002b121c by goroutine 7:
          runtime.slicebytetostringtmp()
              /snap/go/9760/src/runtime/string.go:154 +0x0
          bytes.Equal()
              /snap/go/9760/src/bytes/bytes.go:20 +0x3b5
          github.com/cockroachdb/pebble.(*mergingIter).findNextEntry()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/merging_iter.go:669 +0x2e6
          github.com/cockroachdb/pebble.(*mergingIter).SeekPrefixGE()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/merging_iter.go:938 +0x119
          github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).SeekPrefixGE()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:200 +0xc3
          github.com/cockroachdb/pebble.(*Iterator).SeekPrefixGE()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/iterator.go:1138 +0x9a1
          github.com/cockroachdb/pebble/internal/metamorphic.(*retryableIter).SeekPrefixGE.func1()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/retryable.go:140 +0x76
          github.com/cockroachdb/pebble/internal/metamorphic.(*retryableIter).withRetry()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/retryable.go:39 +0x51
          github.com/cockroachdb/pebble/internal/metamorphic.(*retryableIter).SeekPrefixGE()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/retryable.go:140 +0x94
          github.com/cockroachdb/pebble/internal/metamorphic.(*iterSeekPrefixGEOp).run()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/ops.go:792 +0xa5
          github.com/cockroachdb/pebble/internal/metamorphic.(*test).step()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/test.go:206 +0xea
          github.com/cockroachdb/pebble/internal/metamorphic.testMetaRun()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/meta_test.go:198 +0x132e
          github.com/cockroachdb/pebble/internal/metamorphic.TestMeta()
              /home/arjunnair/go/src/github.com/cockroachdb/pebble/internal/metamorphic/meta_test.go:245 +0xdb3
          testing.tRunner()
              /snap/go/9760/src/testing/testing.go:1439 +0x213
          testing.(*T).Run.func1()
              /snap/go/9760/src/testing/testing.go:1486 +0x47

```

**Comment by bananabrick** [comment]
Option 15 is failing, which has a target file size of 1.

It seems like we're writing and reading to the same byte slice which is the UserKey(verified that it's the UserKey by printing out the pointer addresses).

Read:
https://github.com/cockroachdb/pebble/blob/master/merging_iter.go#L669

Write:
https://github.com/cockroachdb/pebble/blob/master/sstable/block.go#L416

It was taking forever to repro and bisect, so I hacked the metamorphic tests to only run `standard-015` options to find the first commit which is failing.

I was able to bisect it down to: 8e8ff2be5a2fbb57fb79e6eb92305b9809b26a73 and the test doesn't fail on 1b227d2c1513d2f74b09db5e8b8b478912d0b5c1, which is the previous commit.

cc @jbowens. Looking through the commit to see if I can explain this, but you might have a better idea.

**Comment by jbowens** [comment]
Thanks @bananabrick for the bisect. I was unable to find anything suspicious in that commit through examination, so I spent some time tracing the data race bottom up this afternoon.

The data race appears at operation `#5585`. It looks as though the race occurs between a read of one of `iter52`'s `mergingIter`'s level's keys, and a write to a `blockIter` of a compaction's `DebugCheckLevels` iterator.

From what I can tell, it looks as though one of `iter52`'s `SeekPrefixGE` initializes the min heap, loading the sstable that contains the key `uwomy@5`. After initializing the heap, it calls `findNextEntry` to step past keys deleted by range tombstones, which `uwomy@5` is. The level iterator containing the `uwomy@5` sstable is `Next`ed, which closes the sstable iterator and releases the sstable iterator to the sync.Pool.

What comes next is hazy to me. The race seems to be indicating that at some point _after_ we've released the sstable iterator to the pool, the `uwomy@5` key is still set as a `mergingIterItem.key` within the merging iterator, causing us to improperly try to read it.

When the compaction's `DebugCheckLevels` iterator begins iterating, it may retrieve the earlier `singleLevelIterator` from the pool and overwrite the previously read key, triggering the data race.

I need to keep digging to fully understand the read-side of the race and how 8e8ff2be5a2fbb57fb79e6eb92305b9809b26a73 relates.

**Comment by jbowens** [comment]
With [some instrumentation](https://github.com/cockroachdb/pebble/commit/cee350804d6dd5602f049541d144f84bc9ca2564) that tracks `fullKey` buffer pointers of iterators in the pool, I confirmed the `mergingIter` is still holding on to and reading a key from closed sstable reader. In the below execution, iterator `0xc001dab500` with buffer `0xc0002dfb20` is returned to the pool, and the mergingIter reads the buffer `0xc0002dfb20` shortly later. 
 
```
Close returned sstable iterator 0xc001c50e00 with fullKey buffer 0xc00080c528 to the pool.
heapDebugContext seekPrefix after init min heap (mergingIter = 0xc0015c0e20)
github.com/cockroachdb/pebble.(*mergingIter).heapDebugContext:928
github.com/cockroachdb/pebble.(*mergingIter).seekGE:923
github.com/cockroachdb/pebble.(*mergingIter).isNextEntryDeleted:642
github.com/cockroachdb/pebble.(*mergingIter).findNextEntry:662
github.com/cockroachdb/pebble.(*mergingIter).SeekPrefixGE:983

Close returned sstable iterator 0xc0019b3880 with fullKey buffer 0xc001d12fa8 to the pool.
heapDebugContext SeekPrefixGE (mergingIter = 0xc0015c0e20)
github.com/cockroachdb/pebble.(*mergingIter).heapDebugContext:928
runtime.gopanic:844
github.com/cockroachdb/pebble.(*mergingIter).findNextEntry:674
github.com/cockroachdb/pebble.(*mergingIter).SeekPrefixGE:983
github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).SeekPrefixGE:200

Close returned sstable iterator 0xc001dab500 with fullKey buffer 0xc0002dfb20 to the pool.
NewIter acquired sstable iterator 0xc001dab500 with fullKey buffer 0xc0002dfb20
--- FAIL: TestMeta (2.74s)
panic: mergingIter 0xc0015c0e20 attempting to read key 0xc00080c528 ("uwomy@5") [recovered]
	panic: mergingIter 0xc0015c0e20 attempting to read key 0xc00080c528 ("uwomy@5")
```

**Comment by jbowens** [comment]
Put up a fix in #1707. It appears 8e8ff2be5a2fbb57fb79e6eb92305b9809b26a73 is unrelated, maybe just so happening to produce enough of a difference in timing to exercise the race.
